[HttpKernel][FrameworkBundle] Add alternative convention for bundle directories

yceruto opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/issues/32453
| License       | MIT
| Doc PR        | TODO

We already know that bundles must be compatible with many Symfony's versions, so it is very likely that current bundles won't be able to use this feature soon, unless they create symbolic links to support both structures.

The point is that this is already happening, so in the future when our bundles stop to support <=4.3 then you'll be sure to change the current directory structure.

We have recently added the `getPublicDir()` method in https://github.com/symfony/symfony/pull/31975, here I'm removing it in favor of hardcoding a new convention.

I've added some functional tests in which I've changed everything to this structure:
```
-- ModernBundle
   |-- config/ 
   |-- public/
   |-- src/
       |-- ModernBundle.php
   |-- templates/
   |-- translations/
```
WDYT?

yceruto committed:
[HttpKernel][FrameworkBundle] Add alternative convention for bundle directories

nicolas-grekas commented:
As I wrote in https://github.com/symfony/symfony/issues/32453#issuecomment-516491953, I'm not sure we want to make dirs configurable. Even getPublicDir is suspicious to me.

linaori commented:
I think making them configurable isn't necessarily what I want, it's the confusing structure it has right now I'd like to change. Making dirs configurable is one way to do this, because it means I don't have to follow the current structure and that it provides a migration path to change the default.

I very much prefer the structure as shown in the example. The fact that the non-code directory is called "Resources" (mind you, capital R) is bothering me since the start. the "views" directory isn't named correctly anymore either as it's called "templates" in Symfony.

Not sure if the proposed API is a good idea, but I do think that the bundle is responsible for the exposure of certain directories. 
 - "Hey Symfony, you can find templates in this directory!"
 - "Hey, this directory contains entities, if you use doctrine, please register this"
 - etc.

Symfony telling where everything should be feels like something we could change, though it should have sane defaults. I don't agree with the current structure being a sane default anymore due to the changes in the Symfony structure; `app/Resources` vs `bundle/Resources`.

ro0NL commented:
what if we create a single entrypoint similar like `Kernel`, e.g. `$bundle->locateResource(BundlePaths::CONFIG_DIR[, 'validator.yaml'])`

with a built-in deprecation for `views` in favor of `templates`

ro0NL commented on src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/ModernBundle/ModernBundle.php:
```diff
@@ -0,0 +1,29 @@
+
+use Symfony\Component\HttpKernel\Bundle\Bundle;
+
+class ModernBundle extends Bundle
```
shouldnt this technically live in the `src` folder as well? which raises another bigger question; do bundles have a kernel-like `$projectDir`?

yceruto replied:
You're right, it should be in `src`

yceruto replied:
`$bundle->getPath()` is more like `$kernel->getRootDir()`, it's the path where the bundle class/file is placed.

We may need a `$bundle->getBundleDir()` similar to `$kernel->getProjectDir()` assuming the bundle file is in `src`:
```php
public function getBundleDir()
{
    return \dirname(parent::getPath());
}
```
and use `$bundle->getBundleDir().'/public'` to introduce the new conventions, wdyt?

ro0NL replied:
in that case we should deprecate `getPath()` IMHO much like we did with `getRootDir()`

or we actually need it, in that case `getRootDir()` would be in fact the better name

yceruto commented:
> As I wrote in #32453 (comment), I'm not sure we want to make dirs configurable. Even getPublicDir is suspicious to me.

I'm fine if it's not configurable, but we still want a new bundle directory structure, do you want it too?

> with a built-in deprecation for views in favor of templates

@ro0NL we shouldn't deprecating yet -> `... bundles must be compatible with many Symfony's versions.` at least until we're sure we can all migrate to the new structure without any compatibility issue.

> for DI extensions we should also consider injecting the config dir somehow, or access to any path for that matter

I'm not sure it's necessary, you can hardcode it safely there, it's in userland.

nicolas-grekas commented:
> we still want a new bundle directory structure, do you want it too

[yes](https://github.com/orgs/symfony/projects/1#card-24681403)! that'd look consistent with the new directory structure of apps

I'd just do it in `FrameworkExtension` only (and remove/deprecate getPublicDir)

yceruto commented:
**Update**
Removed the configuration of the bundle directory, it is now a new convension that we can optionally use from 4.4 version in the future. 

Added `getBundleDir` method to the Bundle class, which is intended to define the root directory of the bundle. It may differ from the directory where the `Bundle.php` file is located (aka `getPath()`), e.g. when we have the new structure `src/Bundle.php` then `->getBundleDir()` is a folder higher up.

The current convention is not being obsolete yet, it may be in the future.

Convention from 3rd-party bundles like Doctrine ('/Resources/config/doctrine') must progressively support the new method in order to finally be able to deprecating/remove the old structure.

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Bundle/Bundle.php:
```diff
@@ -117,6 +117,11 @@ public function getPath()
         return $this->path;
     }
 
+    public function getBundleDir(): string
```
missing `@inheritdoc`

nicolas-grekas reviewed (approved):
I really like this approach :)
What's the purpose of getPath now? Should we deprecated it? Or should we plan to deprecate it in 5.1, to help bundles migrate smoothly?

yceruto commented:
> What's the purpose of getPath now? Should we deprecated it? Or should we plan to deprecate it in 5.1, to help bundles migrate smoothly?

I think that as of 5.2 (NOV 2020) where we should start to deprecating the legacy structure as well. That'll match with the end of support for 3.4.

yceruto commented:
Sorry for the rebasing after the approved revisions, I needed to update my PhpUnit bridge to make my local phpunit work properly.

Status: This Is Ready In My Side ;)

Tobion commented on src/Symfony/Bundle/FrameworkBundle/Command/AssetsInstallCommand.php:
```diff
@@ -137,13 +137,12 @@ protected function execute(InputInterface $input, OutputInterface $output)
-            if (!method_exists($bundle, 'getPublicDir')) {
-                @trigger_error(sprintf('Not defining "getPublicDir()" method in the "%s" class is deprecated since Symfony 4.4 and will not be supported in 5.0.', \get_class($bundle)), E_USER_DEPRECATED);
-                $publicDir = 'Resources/public';
+            if (method_exists($bundle, 'getBundleDir') && is_dir($bundle->getBundleDir().'/public')) {
```
you are calling `is_dir` twice on the new dir. once here and again in line 145

yceruto replied:
I know, but I don't think it's a big deal, is it? Also, this code is meant to be removed at some point, so it would look like this at end:
```diff
-if (method_exists($bundle, 'getBundleDir') && is_dir($bundle->getBundleDir().'/public')) {
-    $originDir = $bundle->getBundleDir().'/public';
-} else {
-    $originDir = $bundle->getPath().'/Resources/public';
-}
+if (!is_dir($originDir = $bundle->getBundleDir().'/public')) {
    continue;
}
```

yceruto replied:
However, I'd happy to change it with a better alternative that remain it clear enough for the future refactor, do you have one? Thanks!

yceruto replied:
It will change after https://github.com/symfony/symfony/pull/32845#discussion_r311346736 ;)

Tobion commented on src/Symfony/Bundle/FrameworkBundle/Command/AssetsInstallCommand.php:
```diff
@@ -137,13 +137,12 @@ protected function execute(InputInterface $input, OutputInterface $output)
+            if (method_exists($bundle, 'getBundleDir') && is_dir($bundle->getBundleDir().'/public')) {
+                $originDir = $bundle->getBundleDir().'/public';
             } else {
-                $publicDir = ltrim($bundle->getPublicDir(), '\\/');
```
the ltrim path normalization got lost, potentially creating duplicate dir separators

Tobion replied:
If we don't want the normalization, we need to document that `getPublicDir` is expected to return the path without the trailing directory separator. this is enough to me but it should be part of the phpdoc.

yceruto replied:
That method `getPublicDir` doesn't exist in this PR, did you mean `getBundleDir()`?

Like `getPath()`, `getBundleDir()` returns an absolute path managed by `\dirname($reflected->getFileName());` which doesn't add the trailing directory separator, we're confident of it when we use `getPath()` everywhere without check it out.

Still, is it worth mentioning that in the phpdoc block?

Tobion commented:
While I think the new conventions make sense, I'm also of the opinion the directories should not be hardcoded all over the place as you can see in the diff. IMO there should be separate methods like `getTemplateDir`, `getConfigDir` etc. on the BundleInterface. And the default Bundle class implements these new conventions. Even if it's not recommended to change the directories to have conventions across shared bundles, it just makes the discoverability and self-explanation of code much better. It's the same as with the new symfony 4 structure: The default conventions come with the Kernel. Having the default conventions implemented in the Bundle just collects and self-explains those directories that symfony relies on. It should not be spread across all possible places like the assets:install and debug:translations command.

Tobion commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1508,20 +1508,20 @@ private function registerSerializerConfiguration(array $config, ContainerBuilder
 
         foreach ($container->getParameter('kernel.bundles_metadata') as $bundle) {
-            $dirname = $bundle['path'];
+            $configDir = isset($bundle['dir']) && is_dir($bundle['dir'].'/config') ? $bundle['dir'].'/config' : $bundle['path'].'/Resources/config';
```
The legacy dirs should always taken into account. If a bundle used `config` directory in this case for something totally different, the actual `/Resources/config` will be skipped. So an existing bundle stops working just by having a certain directory.

yceruto replied:
Yes, I agree that legacy dirs should have priority. I'll make the changes, thanks!

fabpot commented:
I'm against having methods that allow to change the default directories. A bundle is opinionated and convention based for directories. Developers should be able to easily find things in a third party bundle, so a common directory structure is a must.

yceruto commented:
**Update**

 * Now we'll use the new directory only if the legacy directory doesn't exist.

Status: Needs Review

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Bundle/Bundle.php:
```diff
@@ -117,6 +117,14 @@ public function getPath()
+     */
+    public function getBundleDir(): string
+    {
+        return $this->getPath();
```
if the behavior is exactly the same, we don't need a new method!

yceruto replied:
Indeed, it shouldn't be the default value for the new structure. I've updated it, check it out again.

yceruto commented on src/Symfony/Component/HttpKernel/Bundle/Bundle.php:
```diff
@@ -117,6 +118,22 @@ public function getPath()
+        if (null === $this->dir) {
+            $reflected = new \ReflectionObject($this);
+            $this->dir = \dirname($reflected->getFileName());
+            if (!file_exists($this->dir.'/composer.json') && file_exists(\dirname($this->dir).'/composer.json')) {
```
I guess a dynamic detection of the bundle directory through the `composer.json` existence makes more sense today with bundleless apps approach, though you are free to change it at your convenience.

This new method will be necessary to mix both structures until all our depedencies supporting the new convention thus helping bundles migrate smoothly: 
 * we could move first the `public` dir at the bundle root (try out), 
 * next the `templates` dir (try out) 
 * and later the `translations` dir, etc. 
 * and still have e.g. the `Resources/config` dir under `src/` waiting for another packages that don't support the new convention yet.

yceruto replied:
Another approach would be don't implement the method here and force people to implement it on their own. We're already triggering a deprecation about that. That'll feel safe for us. WDYT?

nicolas-grekas replied:
Or we remove that new method and keep using `getPath`: ppl that want to put the bundle class inside `src/` would `return \dirname(__DIR__);` here

yceruto replied:
Okay, let's move forward with that option, thanks.

yceruto replied:
Done.

nicolas-grekas reviewed (approved):
I love it :)
We shouldn't miss documenting how to put the bundle class inside `src/`

Tobion reviewed (approved):
Well done. Please add a changelog entry (maybe in HttpKernel where the Bundle is defined) explaining the new possible bundle structure. This can then be referred to from the UPGRADE files just mentioning that this is now possible but not required.

yceruto commented:
I missed the last comment, let me know if it's enough.

yceruto commented:
I'm also working on a demo bundle/repo to show people how to migrate safely:
https://github.com/yceruto/acme-bundle (WIP)

nicolas-grekas commented on UPGRADE-4.4.md:
````diff
@@ -108,9 +108,33 @@ HttpFoundation
+    }
+    ```
+
+   As bundles must be compatible with many Symfony versions we are not deprecating
````
`As many bundles must be compatible with a range of Symfony versions,  the current directory convention is not deprecated yet, but it will be in the future.`

maybe duplicate in 5.0 file so that ppl can read about it with new 5.0 projects too

yceruto replied:
done.

nicolas-grekas commented:
Thank you @yceruto.

yceruto commented:
Yujuuu!! :tada:

anyt commented:
Hi,

`Bundle::getPublicDir()` method is [mentioned in the documentation](https://symfony.com/doc/current/bundles.html#overridding-the-bundle-directory-structure) but it's [not used](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Command/AssetsInstallCommand.php#L136). 

Are there any plans to introduce the function back or remove it from documentation?

yceruto commented:
> Are there any plans to introduce the function back or remove it from documentation?

Yes, it should be removed from the documentation (leftover of https://github.com/symfony/symfony/pull/31975)
